compiler/sem/closureiters.nim

```diff
@@ -1291,29 +1284,29 @@ proc deleteEmptyStates(ctx: var Ctx) =
 
   # Apply new state indexes and mark unused states with -1
   var iValid = 0
-  for i, s in ctx.states:
-    let body = skipStmtList(ctx, s[1])
-    if body.kind == nkGotoState and i != ctx.states.len - 1 and i != 0:
+  for i, s in ctx.states.mpairs:
+    let body = skipStmtList(ctx, s.body)
+    if body.kind == nkGotoState and i in 1..(ctx.states.len-2):
       # This is an empty state. Mark with -1.
-      s[0].intVal = -1
+      s.label = -1
     else:
-      s[0].intVal = iValid
+      s.label = iValid
       inc iValid
 
-  for i, s in ctx.states:
-    let body = skipStmtList(ctx, s[1])
+  for i, s in ctx.states.pairs:
+    let body = skipStmtList(ctx, s.body)
     if body.kind != nkGotoState or i == 0:
-      discard ctx.skipThroughEmptyStates(s)
+      discard ctx.skipThroughEmptyStates(s.body)
       let excHandlState = ctx.exceptionTable[i]
       if excHandlState < 0:
         ctx.exceptionTable[i] = -ctx.skipEmptyStates(-excHandlState)
       elif excHandlState != 0:
         ctx.exceptionTable[i] = ctx.skipEmptyStates(excHandlState)
 
-  var i = 0
+  # remove unused states except for the first (entry) and last one (exit)
+  var i = 1
   while i < ctx.states.len - 1:
-    let fs = skipStmtList(ctx, ctx.states[i][1])
-    if fs.kind == nkGotoState and i != 0:
+    if ctx.states[i].label == -1: # is it unused?
```

Comment:
Very soft suggestion: make `-1` a constant, such as `unusedStateIdx`, might read a bit better. It'd also be applicable to the setting of the label values above.

Reply:
That's much better, thanks.